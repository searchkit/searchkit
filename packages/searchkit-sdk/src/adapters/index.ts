import { SearchkitConfig } from '..'

export interface SearchkitAdapterConstructor {
  new (config: SearchkitConfig): SearchkitAdapter
}

export interface SearchkitAdapter {
  performRequest(requestBody): Promise<any>
}
