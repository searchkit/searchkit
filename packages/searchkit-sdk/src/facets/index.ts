export { default as RefinementSelectFacet } from './RefinementSelectFacet'
export { default as RangeFacet } from './RangeFacet'
export { default as DateRangeFacet } from './DateRangeFacet'
export { default as HierarchicalMenuFacet } from './HierarchicalMenuFacet'
export { default as MultiQueryOptionsFacet } from './MultiQueryOptionsFacet'
export { default as GeoDistanceOptionsFacet } from './GeoDistanceOptionsFacet'
export * from './BaseFacet'
export * from './VisibilityRules'
