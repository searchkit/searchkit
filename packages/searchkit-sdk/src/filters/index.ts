export { default as GeoBoundingBoxFilter } from './Filter'
export { default as TermFilter } from './Filter'
export { default as NumericRangeFilter } from './Filter'
export { default as Filter } from './Filter'
export * from './BaseFilter'
