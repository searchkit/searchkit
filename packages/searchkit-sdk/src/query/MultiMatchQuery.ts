import QueryManager from '../core/QueryManager'
import BaseQuery from './BaseQuery'

interface MultiMatchQueryConfig {
  fields: string[]
}

class MultiMatchQuery implements BaseQuery {
  constructor(private config: MultiMatchQueryConfig) {}

  getFilter(queryManager: QueryManager) {
    if (queryManager.hasQuery()) {
      return {
        bool: {
          should: [
            {
              multi_match: {
                query: queryManager.getQuery(),
                fields: this.config.fields,
                type: 'best_fields',
                operator: 'and'
              }
            },
            {
              multi_match: {
                query: queryManager.getQuery(),
                fields: this.config.fields,
                type: 'cross_fields'
              }
            },
            {
              multi_match: {
                query: queryManager.getQuery(),
                fields: this.config.fields,
                type: 'phrase'
              }
            },
            {
              multi_match: {
                query: queryManager.getQuery(),
                fields: this.config.fields,
                type: 'phrase_prefix'
              }
            }
          ]
        }
      }
    }
    return null
  }
}

export default MultiMatchQuery
