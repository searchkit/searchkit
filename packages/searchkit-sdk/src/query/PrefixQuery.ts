import QueryManager from '../core/QueryManager'
import { Query } from '../core/RequestBodyBuilder'
import BaseQuery from './BaseQuery'

interface PrefixQueryOptions {
  fields: string[]
}

export default class PrefixQuery implements BaseQuery {
  private fields: string[]

  constructor(options: PrefixQueryOptions) {
    this.fields = options.fields
  }

  getFilter(queryManager: QueryManager): Query {
    return {
      bool: {
        must: [
          {
            multi_match: {
              query: queryManager.getQuery(),
              type: 'bool_prefix',
              fields: this.fields
            }
          }
        ]
      }
    }
  }
}
