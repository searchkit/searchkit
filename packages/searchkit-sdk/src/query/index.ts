export { default as MultiMatchQuery } from './MultiMatchQuery'
export { default as BaseQuery } from './BaseQuery'
export { default as CustomQuery } from './CustomQuery'
export { default as PrefixQuery } from './PrefixQuery'
