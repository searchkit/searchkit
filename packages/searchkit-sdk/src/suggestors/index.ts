import { SearchkitTransporter } from '../transporters'

export interface BaseSuggestor<T> {
  getSuggestions: (query: string, transport: SearchkitTransporter) => Promise<T>
}

export interface BaseSuggestorResponse {
  identifier: string
}

export * from './HitsSuggestor'
export * from './CompletionSuggestor'
