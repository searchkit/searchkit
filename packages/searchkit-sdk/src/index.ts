import type { SearchInnerHits, SearchRequest } from '@elastic/elasticsearch-types/lib/api/types'
import QueryManager, { MixedFilter, QueryOptions } from './core/QueryManager'
import RequestBodyBuilder from './core/RequestBodyBuilder'
import BaseQuery from './query/BaseQuery'
import { BaseFacet } from './facets/BaseFacet'
import { BaseFilter } from './filters/BaseFilter'
import { VisibleWhenRuleSet } from './facets'
import { SearchkitTransporter, FetchClientTransporter } from './transporters'
import { getAggregationsFromFacets } from './core/FacetsFns'
import {
  SearchkitResponseTransformer,
  ElasticSearchResponseTransformer,
  SearchkitResponse,
  SearchkitHit
} from './transformers'
import { BaseSuggestor } from './suggestors'

export * from './query'
export * from './facets'
export * from './filters'
export * from './transporters'
export * from './suggestors'
export { MixedFilter, QueryOptions, SearchkitResponse, SearchkitHit }

export interface SortingOption {
  id: string
  label: string
  field: any
  defaultOption?: boolean
}

export interface CustomHighlightConfig {
  field: string
  config: any
}

export interface CloudHost {
  id: string
}

export interface BaseConfig {
  host?: string
  cloud?: CloudHost
  index: string
  connectionOptions?: {
    apiKey?: string
    headers?: Record<string, string>
  }
}

export type CollapseInnerHitsConfig = SearchInnerHits

export interface CollapseConfig {
  field: string
  inner_hits?: CollapseInnerHitsConfig | CollapseInnerHitsConfig[]
}

export interface SearchkitConfig extends BaseConfig {
  suggestions?: Array<BaseSuggestor<any>>
  sortOptions?: SortingOption[]
  hits?: {
    fields: string[]
    highlightedFields?: (string | CustomHighlightConfig)[]
  }
  query?: BaseQuery
  facets?: Array<BaseFacet | VisibleWhenRuleSet>
  filters?: Array<BaseFilter>
  collapse?: CollapseConfig
  postProcessRequest?: (body: SearchRequest) => SearchRequest
}

export interface ResultsResolverParameters {
  filters: Array<MixedFilter>
  query: string
}

const getFacets = (
  facets: Array<BaseFacet | VisibleWhenRuleSet> = [],
  queryManager: QueryManager,
  ctx
) =>
  facets.reduce((facetsList, facet) => {
    if (facet instanceof VisibleWhenRuleSet) {
      return [...facetsList, ...facet.getActiveFacets(queryManager, ctx)]
    }
    return [...facetsList, facet]
  }, [])

const createInstance = (
  config: SearchkitConfig,
  transporter?: SearchkitTransporter
): SearchkitRequest => new SearchkitRequest(config, transporter)

export type ResponseRequest = {
  hits?: {
    size?: number
    from?: number
    includeRawHit?: boolean
  }
  facets?:
    | boolean
    | Array<{
        identifier: string
        query?: string
        size?: number
      }>
}

export type BaseFilters = Array<any>

function getSortOption(id, sortOptions: SortingOption[]) {
  const selectedSortOption = sortOptions.find((sortOption) => sortOption.id === id)
  if (!selectedSortOption) {
    throw new Error(`Sort Option: sorting option ${id} not found`)
  }
  return selectedSortOption
}

export class SearchkitRequest {
  private queryManager: QueryManager
  private transporter: SearchkitTransporter
  private transformer: SearchkitResponseTransformer

  constructor(private config: SearchkitConfig, transporter?: SearchkitTransporter) {
    this.queryManager = new QueryManager()
    this.transporter = !transporter ? new FetchClientTransporter(config) : transporter
    this.transformer = new ElasticSearchResponseTransformer()
    this.handleDefaults()
  }

  handleDefaults(): void {
    const defaultSortOptionId =
      this.config.sortOptions && this.config.sortOptions.find((option) => option.defaultOption)
    const defaultSortOption =
      defaultSortOptionId && getSortOption(defaultSortOptionId.id, this.config.sortOptions)
    this.queryManager.setSortBy(defaultSortOption)
  }

  query(query: string): SearchkitRequest {
    this.queryManager.setQuery(query)
    return this
  }

  setFilters(filters: Array<MixedFilter>): SearchkitRequest {
    this.queryManager.setFilters(filters)
    return this
  }

  setQueryOptions(options: QueryOptions): SearchkitRequest {
    this.queryManager.setQueryOptions(options)
    return this
  }

  setSortBy(sortBy: string): SearchkitRequest {
    if (this.config.sortOptions && sortBy) {
      const sortOption = getSortOption(sortBy, this.config.sortOptions)
      this.queryManager.setSortBy(sortOption)
    }
    return this
  }

  async executeSuggestions(query: string): Promise<Array<any>> {
    const suggestions = await Promise.all(
      this.config.suggestions.map((suggestor) => suggestor.getSuggestions(query, this.transporter))
    )
    return suggestions
  }

  async execute(
    responseRequest: ResponseRequest,
    baseFilters: BaseFilters = []
  ): Promise<SearchkitResponse> {
    const partialQueries = []
    const facets = getFacets(this.config.facets, this.queryManager, {})
    let filteredFacets = null
    let overrides = {}

    if (!responseRequest.hits) responseRequest.hits = {}
    if (!responseRequest.hits.size) responseRequest.hits.size = 0
    if (!responseRequest.hits.from) responseRequest.hits.from = 0

    if (responseRequest.facets) {
      if (Array.isArray(responseRequest.facets)) {
        const filteredFacetIdentifiers = responseRequest.facets.map((facet) => facet.identifier)
        filteredFacets =
          facets &&
          facets.filter((facet) => filteredFacetIdentifiers.includes(facet.getIdentifier()))
        overrides = responseRequest.facets.reduce(
          (acc, facet) => ({
            ...acc,
            [facet.identifier]: {
              query: facet.query,
              size: facet.size
            }
          }),
          {}
        )
      }
      const aggs = getAggregationsFromFacets(this.queryManager, overrides, filteredFacets || facets)
      partialQueries.push(aggs)
    }

    const chosenSortOption = this.queryManager.getSortBy()

    partialQueries.push({
      size: responseRequest.hits.size,
      from: responseRequest.hits.from,
      sort: chosenSortOption ? chosenSortOption.field : [{ _score: 'desc' }]
    })

    let skRequestBody = RequestBodyBuilder(
      this.queryManager,
      this.config,
      baseFilters,
      facets,
      partialQueries
    )
    if (this.config.postProcessRequest) {
      skRequestBody = this.config.postProcessRequest(skRequestBody)
    }
    const response = await this.transporter.performRequest(skRequestBody)

    return this.transformer.transformResponse(
      response,
      filteredFacets || facets,
      this.queryManager,
      this.config,
      responseRequest
    )
  }
}

export default createInstance
