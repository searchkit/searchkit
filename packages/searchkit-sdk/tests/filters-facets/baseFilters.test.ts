import nock from 'nock'
import SearchkitRequest from '../../src'
import { MultiMatchQuery } from '../../src/query'
import HitsMMock from '../__mock-data__/HitResolver/Hits.json'

describe('Base Filters', () => {
  it('Example implementation', async () => {
    const request = SearchkitRequest({
      host: 'http://localhost:9200',
      query: new MultiMatchQuery({
        fields: ['title', 'body']
      }),
      hits: {
        fields: ['facet1']
      },
      index: 'test'
    })

    request.query('test')

    const scope = nock('http://localhost:9200')
      .post('/test/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchInlineSnapshot(`
          Object {
            "_source": Object {
              "includes": Array [
                "facet1",
              ],
            },
            "aggs": Object {},
            "from": 0,
            "query": Object {
              "bool": Object {
                "filter": Array [
                  Object {
                    "term": Object {
                      "status": "published",
                    },
                  },
                ],
                "should": Array [
                  Object {
                    "multi_match": Object {
                      "fields": Array [
                        "title",
                        "body",
                      ],
                      "operator": "and",
                      "query": "test",
                      "type": "best_fields",
                    },
                  },
                  Object {
                    "multi_match": Object {
                      "fields": Array [
                        "title",
                        "body",
                      ],
                      "query": "test",
                      "type": "cross_fields",
                    },
                  },
                  Object {
                    "multi_match": Object {
                      "fields": Array [
                        "title",
                        "body",
                      ],
                      "query": "test",
                      "type": "phrase",
                    },
                  },
                  Object {
                    "multi_match": Object {
                      "fields": Array [
                        "title",
                        "body",
                      ],
                      "query": "test",
                      "type": "phrase_prefix",
                    },
                  },
                ],
              },
            },
            "size": 10,
            "sort": Array [
              Object {
                "_score": "desc",
              },
            ],
          }
        `)
        expect((body as Record<string, any>).size).toBe(10)
        return HitsMMock
      })

    const response = await request.execute(
      {
        hits: {
          size: 10
        }
      },
      [{ term: { status: 'published' } }]
    )
    expect(response).toMatchSnapshot()
    expect(response.facets).toBeFalsy()
    expect(response.summary.query).toBe('test')
    expect(response.summary.total).toBe(4162)
  })
})
