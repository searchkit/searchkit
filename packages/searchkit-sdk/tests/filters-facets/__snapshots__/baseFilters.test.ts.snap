// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base Filters Example implementation 2`] = `
Object {
  "facets": null,
  "hits": Object {
    "items": Array [
      Object {
        "fields": Object {
          "actors": Array [
            "Christian Bale",
            "Heath Ledger",
            "Aaron Eckhart",
            "Michael Caine",
          ],
          "awards": "Won 2 Oscars. Another 94 wins & 69 nominations.",
          "countries": Array [
            "USA",
            "UK",
          ],
          "genres": Array [
            "Action",
            "Crime",
            "Drama",
          ],
          "imdbId": "tt0468569",
          "imdbRating": 9,
          "imdbVotes": 1133701,
          "metaScore": 82,
          "plot": "When Batman, Gordon and Harvey Dent launch an assault on the mob, they let the clown out of the box, the Joker, bent on turning Gotham on itself and bringing any heroes down to his level.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BMTMxNTMwODM0NF5BMl5BanBnXkFtZTcwODAyMTk2Mw@@._V1_SX300.jpg",
          "rated": Array [
            "PG-13",
          ],
          "released": "2008-07-18",
          "runtimeMinutes": 152,
          "title": "The Dark Knight",
          "type": "Movie",
          "writers": Array [
            "Jonathan Nolan",
            "Christopher Nolan",
            "David S. Goyer",
            "Bob Kane",
          ],
          "year": 2008,
        },
        "highlight": Object {},
        "id": "tt0468569",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Tim Roth",
            "Amanda Plummer",
            "Laura Lovelace",
            "John Travolta",
          ],
          "awards": "Won 1 Oscar. Another 61 wins & 46 nominations.",
          "countries": Array [
            "USA",
          ],
          "genres": Array [
            "Crime",
            "Drama",
            "Thriller",
          ],
          "imdbId": "tt0110912",
          "imdbRating": 9,
          "imdbVotes": 895306,
          "metaScore": 94,
          "plot": "The lives of two mob hit men, a boxer, a gangster's wife, and a pair of diner bandits intertwine in four tales of violence and redemption.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BMjE0ODk2NjczOV5BMl5BanBnXkFtZTYwNDQ0NDg4._V1_SX300.jpg",
          "rated": Array [
            "R",
          ],
          "released": "1994-10-14",
          "runtimeMinutes": 154,
          "title": "Pulp Fiction",
          "type": "Movie",
          "writers": Array [
            "Quentin Tarantino",
            "Roger Avary",
          ],
          "year": 1994,
        },
        "highlight": Object {},
        "id": "tt0110912",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Noel Appleby",
            "Alexandra Astin",
            "Sean Astin",
            "David Aston",
          ],
          "awards": "Won 11 Oscars. Another 122 wins & 73 nominations.",
          "countries": Array [
            "USA",
            "New Zealand",
          ],
          "genres": Array [
            "Action",
            "Adventure",
            "Fantasy",
          ],
          "imdbId": "tt0167260",
          "imdbRating": 8.9,
          "imdbVotes": 826369,
          "metaScore": 94,
          "plot": "Gandalf and Aragorn lead the World of Men against Sauron's army to draw his gaze from Frodo and Sam as they approach Mount Doom with the One Ring.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BMjE4MjA1NTAyMV5BMl5BanBnXkFtZTcwNzM1NDQyMQ@@._V1_SX300.jpg",
          "rated": Array [
            "PG-13",
          ],
          "released": "2003-12-17",
          "runtimeMinutes": 201,
          "title": "The Lord of the Rings: The Return of the King",
          "type": "Movie",
          "writers": Array [
            "J.R.R. Tolkien",
            "Fran Walsh",
            "Philippa Boyens",
            "Peter Jackson",
          ],
          "year": 2003,
        },
        "highlight": Object {},
        "id": "tt0167260",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Morgan Freeman",
            "Andrew Kevin Walker",
            "Daniel Zacapa",
            "Brad Pitt",
          ],
          "awards": "Nominated for 1 Oscar. Another 25 wins & 18 nominations.",
          "countries": Array [
            "USA",
          ],
          "genres": Array [
            "Crime",
            "Mystery",
            "Thriller",
          ],
          "imdbId": "tt0114369",
          "imdbRating": 8.7,
          "imdbVotes": 672372,
          "metaScore": 65,
          "plot": "Two detectives, a rookie and a veteran, hunt a serial killer who uses the seven deadly sins as his modus operandi.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BMTQwNTU3MTE4NF5BMl5BanBnXkFtZTcwOTgxNDM2Mg@@._V1_SX300.jpg",
          "rated": Array [
            "R",
          ],
          "released": "1995-09-22",
          "runtimeMinutes": 127,
          "title": "Se7en",
          "type": "Movie",
          "writers": Array [
            "Andrew Kevin Walker",
          ],
          "year": 1995,
        },
        "highlight": Object {},
        "id": "tt0114369",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Michael Berryman",
            "Peter Brocco",
            "Dean R. Brooks",
            "Alonzo Brown",
          ],
          "awards": "Won 5 Oscars. Another 30 wins & 12 nominations.",
          "countries": Array [
            "USA",
          ],
          "genres": Array [
            "Drama",
          ],
          "imdbId": "tt0073486",
          "imdbRating": 8.8,
          "imdbVotes": 481493,
          "metaScore": 79,
          "plot": "Upon admittance to a mental institution, a brash rebel rallies the patients to take on the oppressive head nurse.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BMTk5OTA4NTc0NF5BMl5BanBnXkFtZTcwNzI5Mzg3OA@@._V1_SX300.jpg",
          "rated": Array [
            "R",
          ],
          "released": "1975-11-21",
          "runtimeMinutes": 133,
          "title": "One Flew Over the Cuckoo's Nest",
          "type": "Movie",
          "writers": Array [
            "Lawrence Hauben",
            "Bo Goldman",
            "Ken Kesey",
            "Dale Wasserman",
          ],
          "year": 1975,
        },
        "highlight": Object {},
        "id": "tt0073486",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Russell Crowe",
            "Joaquin Phoenix",
            "Connie Nielsen",
            "Oliver Reed",
          ],
          "awards": "Won 5 Oscars. Another 51 wins & 80 nominations.",
          "countries": Array [
            "USA",
            "UK",
          ],
          "genres": Array [
            "Action",
            "Adventure",
            "Drama",
          ],
          "imdbId": "tt0172495",
          "imdbRating": 8.5,
          "imdbVotes": 649422,
          "metaScore": 64,
          "plot": "When a Roman general is betrayed and his family murdered by an emperor's corrupt son, he comes to Rome as a gladiator to seek revenge.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BMTgwMzQzNTQ1Ml5BMl5BanBnXkFtZTgwMDY2NTYxMTE@._V1_SX300.jpg",
          "rated": Array [
            "R",
          ],
          "released": "2000-05-05",
          "runtimeMinutes": 155,
          "title": "Gladiator",
          "type": "Movie",
          "writers": Array [
            "David Franzoni",
            "John Logan",
            "William Nicholson",
          ],
          "year": 2000,
        },
        "highlight": Object {},
        "id": "tt0172495",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Andrew Lincoln",
            "Steven Yeun",
            "Chandler Riggs",
            "Norman Reedus",
          ],
          "awards": "Nominated for 1 Golden Globe. Another 15 wins & 50 nominations.",
          "countries": Array [
            "USA",
          ],
          "genres": Array [
            "Drama",
            "Horror",
            "Thriller",
          ],
          "imdbId": "tt1520211",
          "imdbRating": 8.8,
          "imdbVotes": 357942,
          "plot": "Police officer Rick Grimes leads a group of survivors in a world overrun by zombies.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BNzU0MjIzNzY1MV5BMl5BanBnXkFtZTgwNjgxNTg4MDE@._V1_SX300.jpg",
          "rated": Array [
            "TV-14",
          ],
          "released": "2010-10-31",
          "runtimeMinutes": 45,
          "title": "The Walking Dead",
          "type": "Series",
          "writers": Array [
            "Frank Darabont",
          ],
          "year": 2010,
        },
        "highlight": Object {},
        "id": "tt1520211",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Jim Carrey",
            "Kate Winslet",
            "Gerry Robert Byrne",
            "Elijah Wood",
          ],
          "awards": "Won 1 Oscar. Another 46 wins & 54 nominations.",
          "countries": Array [
            "USA",
          ],
          "genres": Array [
            "Drama",
            "Romance",
            "Sci-Fi",
          ],
          "imdbId": "tt0338013",
          "imdbRating": 8.4,
          "imdbVotes": 467149,
          "metaScore": 89,
          "plot": "A couple undergo a procedure to erase each other from their memories when their relationship turns sour, but it is only through the process of loss that they discover what they had to begin with.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BMTY4NzcwODg3Nl5BMl5BanBnXkFtZTcwNTEwOTMyMw@@._V1_SX300.jpg",
          "rated": Array [
            "R",
          ],
          "released": "2004-03-19",
          "runtimeMinutes": 108,
          "title": "Eternal Sunshine of the Spotless Mind",
          "type": "Movie",
          "writers": Array [
            "Charlie Kaufman",
            "Michel Gondry",
            "Pierre Bismuth",
          ],
          "year": 2004,
        },
        "highlight": Object {},
        "id": "tt0338013",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Marlon Brando",
            "Martin Sheen",
            "Robert Duvall",
            "Frederic Forrest",
          ],
          "awards": "Won 2 Oscars. Another 17 wins & 30 nominations.",
          "countries": Array [
            "USA",
          ],
          "genres": Array [
            "Drama",
            "War",
          ],
          "imdbId": "tt0078788",
          "imdbRating": 8.5,
          "imdbVotes": 329282,
          "metaScore": 90,
          "plot": "During the U.S.-Viet Nam War, Captain Willard is sent on a dangerous mission into Cambodia to assassinate a renegade colonel who has set himself up as a god among a local tribe.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BMTcyMzQ5NDM4OV5BMl5BanBnXkFtZTcwODUwNDg3OA@@._V1_SX300.jpg",
          "rated": Array [
            "R",
          ],
          "released": "1979-08-15",
          "runtimeMinutes": 153,
          "title": "Apocalypse Now",
          "type": "Movie",
          "writers": Array [
            "John Milius",
            "Francis Ford Coppola",
            "Michael Herr",
          ],
          "year": 1979,
        },
        "highlight": Object {},
        "id": "tt0078788",
      },
      Object {
        "fields": Object {
          "actors": Array [
            "Ellen Burstyn",
            "Jared Leto",
            "Jennifer Connelly",
            "Marlon Wayans",
          ],
          "awards": "Nominated for 1 Oscar. Another 27 wins & 36 nominations.",
          "countries": Array [
            "USA",
          ],
          "genres": Array [
            "Drama",
          ],
          "imdbId": "tt0180093",
          "imdbRating": 8.4,
          "imdbVotes": 406824,
          "metaScore": 68,
          "plot": "The drug-induced utopias of four Coney Island individuals are shattered when their addictions become stronger.",
          "poster": "https://s3-eu-west-1.amazonaws.com/imdbimages/images/MV5BMzM2OTYwMTY4Nl5BMl5BanBnXkFtZTcwMjU1Njg3OA@@._V1_SX300.jpg",
          "rated": Array [
            "R",
          ],
          "released": "2000-10-27",
          "runtimeMinutes": 102,
          "title": "Requiem for a Dream",
          "type": "Movie",
          "writers": Array [
            "Hubert Selby Jr.",
            "Darren Aronofsky",
          ],
          "year": 2000,
        },
        "highlight": Object {},
        "id": "tt0180093",
      },
    ],
    "page": Object {
      "from": 0,
      "pageNumber": 0,
      "size": 10,
      "total": 4162,
      "totalPages": 417,
    },
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [],
    "disabledFilters": Array [],
    "query": "test",
    "sortOptions": Array [],
    "total": 4162,
  },
}
`;
